sklearn/utils/extmath.py

```diff
@@ -472,13 +508,19 @@ def randomized_svd(
     )
 
     # project M to the (k + p) dimensional space using the basis vectors
-    B = safe_sparse_dot(Q.T, M)
+    B = Q.T @ M
 
     # compute the SVD on the thin matrix: (k + p) wide
-    Uhat, s, Vt = linalg.svd(B, full_matrices=False, lapack_driver=svd_lapack_driver)
+    xp, is_array_api_compliant = get_namespace(B)
+    if _is_numpy_namespace(xp) and not is_array_api_compliant:
+        Uhat, s, Vt = linalg.svd(
+            B, full_matrices=False, lapack_driver=svd_lapack_driver
+        )
+    else:
+        Uhat, s, Vt = xp.linalg.svd(B, full_matrices=False)
```

Comment:
@thomasjpfan I have the feeling that `_is_numpy_namespace(xp)` is redundant here. Do you envision a future where `is_array_api_compliant is False` and `xp.__namespace__ != "numpy"`?

Otherwise:


```suggestion
    if is_array_api_compliant:
        Uhat, s, Vt = xp.linalg.svd(B, full_matrices=False)
    else:
        # Rely on scipy.linalg instead of numpy.linalg to avoid introducing
        # a behavior change with previous versions of scikit-learn when
        # array_api_dispatch is False.
        Uhat, s, Vt = linalg.svd(
            B, full_matrices=False, lapack_driver=svd_lapack_driver
        )
```

Reply:
We should add a comment for people from the future. The way I understand it is that if the input is Numpy or Array API support is turned off we want to use the `svd` from scipy. So that the results don't change compared to what is in scikit-learn now.

In the other cases (not a Numpy array and Array API support turned on) we want to use the `svd` from `xp.linalg`.

But I think neither of the proposed bits of code make that obvious. They all check something that is highly correlated with this, but not exactly this. Hence leaving a comment for people from the future.

Reply:
> if the input is Numpy or Array API support is turned off we want to use the svd from scipy.

Not exactly: at the moment, if `xp.__namespace__ == "numpy"` (but not `xp.__namespace__ == "numpy.array_api"`) and Array API dispatch is turned off, then we use scipy. Otherwise we use `xp.linalg` (e.g. `numpy.array_api.linalg` when `xp.__namespace__ == "numpy.array_api"`).

Reply:
BTW: I added the comment in the last commit: e30bfa8 (without changing the condition).